### ISSUE TITLE
Failed to read response when sending data

### ISSUE THREAD

**Issue by dpeters-ipc** [adjust/android_sdk#228]
Environment:
---
Adjust Version: 4.10.1 (Sandbox-Mode)
Device: Samsung Galaxy S4
Device-Android Version: 5.0.1

Since Adjust 4.10.1 I get this error in logcat for every event:

```
E  Failed to read response. (Attempt to invoke virtual method 'com.android.okhttp.Route com.android.okhttp.Connection.getRoute()' on a null object reference)
E  Failed to track session. (Runtime exception: java.lang.NullPointerException: Attempt to invoke virtual method 'com.android.okhttp.Route com.android.okhttp.Connection.getRoute()' on a null object reference)
```

In 4.7.0 the error wasn't thrown, instead I see some debug json-data.
On another device with Android 7.0 it works also with 4.10.1, so I would think that it is a SDK related issue.

Since I cannot be sure if everything is tracked correctly, I reverted to 4.7.0 again.

Below are debug-logs from both Adjust versions after first app-start:

4.10.1:
---
```
Adjust  F  SANDBOX: Adjust is running in Sandbox mode. Use this setting for testing. Don't forget to set the environment to `production` before publishing!
                         E  Adjust not initialized correctly
                         D  Attribution file not found
                         D  Activity state file not found
                         D  Session Callback parameters file not found
                         D  Session Partner parameters file not found
                         I  Google Play Services Advertising ID read correctly at start time
                         V  Foreground timer configured to fire after 60,0 seconds of starting and cycles every 60,0 seconds
                         D  Package queue file not found
                         V  Foreground timer starting
                         V  Subsession start
                         D  Added package 1 (session)
                         D  Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:xxxxx
                         V  Path:      /session
                         V  ClientSdk: android4.10.1
                         V  Parameters:
                         V      android_uuid     xxxxx
                         V      api_level        21
                         V      app_token        xxxxx
                         V      app_version      xxxxx
                         V      attribution_deeplink 1
                         V      country          DE
                         V      cpu_type         armeabi-v7a
                         V      created_at       2016-09-15T16:47:03.882Z+0200
                         V      device_manufacturer samsung
                         V      device_name      GT-I9505
                         V      device_type      phone
                         V      display_height   1920
                         V      display_width    1080
                         V      environment      sandbox
                         V      event_buffering_enabled 0
                         V      gps_adid         xxxxx
                         V      hardware_name    LRX22C.I9505XXUHOJ2
                         V      language         de
                         V      needs_response_details 0
                         V      os_name          android
                         V      os_version       5.0.1
                         V      package_name     xxxxx
                         V      screen_density   high
                         V      screen_format    long
                         V      screen_size      normal
                         V      session_count    1
                         V      tracking_enabled 0
                         D  Wrote Package queue: [session]
                         D  Package handler wrote 1 packages
                         E  Failed to read response. (Attempt to invoke virtual method 'com.android.okhttp.Route com.android.okhttp.Connection.getRoute()' on a null object reference)
                         E  Failed to track session. (Runtime exception: java.lang.NullPointerException: Attempt to invoke virtual method 'com.android.okhttp.Route com.android.okhttp.Connection.getRoute()' on a null object refe
                            rence)
                         D  Wrote Package queue: []
                         D  Package handler wrote 0 packages
                         V  Package handler can send
```

4.7.0:
---
```
Adjust  F  SANDBOX: Adjust is running in Sandbox mode. Use this setting for testing. Don't forget to set the environment to `production` before publishing!
                         E  Adjust not initialized correctly
                         D  Attribution file not found
                         D  Activity state file not found
                         I  Google Play Services Advertising ID read correctly at start time
                         D  Package queue file not found
                         V  Background timer canceled
                         V  Foreground timer starting in 60,0 seconds and cycle every 60,0 seconds
                         V  Subsession start
                         D  Added package 1 (session)
                         V  Path:      /session
                         V  ClientSdk: android4.7.0
                         V  Parameters:
                         V      android_uuid     xxxxx
                         V      api_level        21
                         V      app_token        xxxxx
                         V      app_version      xxxxx
                         V      country          DE
                         V      cpu_type         armeabi-v7a
                         V      created_at       2016-09-15T16:39:08.141Z+0200
                         V      device_manufacturer samsung
                         V      device_name      GT-I9505
                         V      device_type      phone
                         V      display_height   1920
                         V      display_width    1080
                         V      environment      sandbox
                         V      event_buffering_enabled 0
                         V      gps_adid         xxxxx
                         V      hardware_name    LRX22C.I9505XXUHOJ2
                         V      language         de
                         V      needs_response_details 0
                         V      os_name          android
                         V      os_version       5.0.1
                         V      package_name     xxxxx
                         V      screen_density   high
                         V      screen_format    long
                         V      screen_size      normal
                         V      session_count    1
                         V      tracking_enabled 0
                         D  Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:dab4a353-75e5-4c61-8b84-06d9a356d569
                         D  Wrote Package queue: [session]
                         D  Package handler wrote 1 packages
                         V  Response: {"app_token":"xxxxx","adid":"xxxxx","timestamp":"2016-09-15T14:39:09.070Z+0000","message":"Session tracked","attribution":{"tracker_token":"xxxxx","tracke
                            r_name":"Organic","network":"Organic"}}
                         I  Session tracked
                         D  Wrote Package queue: []
                         D  Package handler wrote 0 packages
                         V  Package handler can send
                         D  Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:xxxxx
                         D  Wrote Attribution: tt:aem10f tn:Organic net:Organic cam:null adg:null cre:null cl:null
```

**Comment by uerceg** [comment]
Hi @dpeters-ipc 

Are you able to reproduce this issue all the time or is it randomly appearing? Are you experiencing it only with this device or it is device unrelated?

Cheers

**Comment by dpeters-ipc** [comment]
On this device always and reproducable. I am currently checking if I can find another affected device.

**Comment by dpeters-ipc** [comment]
I can reproduce it in an Android Emulator with API 21 (5.0.2), too:

```
 Adjust  F  SANDBOX: Adjust is running in Sandbox mode. Use this setting for testing. Don't forget to set the environment to `production` before publishing!
                         E  Adjust not initialized correctly
                         D  Attribution file not found
                         D  Activity state file not found
                         D  Session Callback parameters file not found
                         D  Session Partner parameters file not found
                         I  Google Play Services Advertising ID read correctly at start time
                         V  Foreground timer configured to fire after 60.0 seconds of starting and cycles every 60.0 seconds
                         D  Package queue file not found
                         V  Foreground timer starting
                         V  Subsession start
                         D  Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:xxxxx
                         D  Added package 1 (session)
                         V  Path:      /session
                         V  ClientSdk: android4.10.1
                         V  Parameters:
                         V      android_uuid     xxxxx
                         V      api_level        21
                         V      app_token        xxxxx
                         V      app_version      xxxxx
                         V      attribution_deeplink 1
                         V      country          GB
                         V      cpu_type         x86_64
                         V      created_at       2016-09-15T18:33:18.031Z+0200
                         V      device_manufacturer unknown
                         V      device_name      Android SDK built for x86_64
                         V      device_type      phone
                         V      display_height   1794
                         V      display_width    1080
                         V      environment      sandbox
                         V      event_buffering_enabled 0
                         V      gps_adid         xxxxx
                         V      hardware_name    sdk_google_phone_x86_64-eng 5.0.2 LSY66K 3079185 test-keys
                         V      language         en
                         V      needs_response_details 0
                         V      os_name          android
                         V      os_version       5.0.2
                         V      package_name     xxxxx
                         V      screen_density   high
                         V      screen_format    normal
                         V      screen_size      normal
                         V      session_count    1
                         V      tracking_enabled 1
                         D  Wrote Package queue: [session]
                         D  Package handler wrote 1 packages
                         E  Failed to read response. (Attempt to invoke virtual method 'com.android.okhttp.Route com.android.okhttp.Connection.getRoute()' on a null object reference)
                         E  Failed to track session. (Runtime exception: java.lang.NullPointerException: Attempt to invoke virtual method 'com.android.okhttp.Route com.android.okhttp.Connection.getRoute()' on a null object refe
                            rence)
                         D  Wrote Package queue: []
                         D  Package handler wrote 0 packages
                         V  Package handler can send
```

**Comment by dpeters-ipc** [comment]
the problem also exists on an HTC One Max (5.0.1)

**Comment by uerceg** [comment]
Thanks for the update, we're on it.

**Comment by nonelse** [comment]
HttpsURLConnection is being disconnected before the response is read.
Change done in [PR#229](https://github.com/adjust/android_sdk/pull/229) to be released very soon

**Comment by nonelse** [comment]
Fixed in [v4.10.2](https://github.com/adjust/android_sdk/releases/tag/v4.10.2)

**Comment by dpeters-ipc** [comment]
thanks for the very fast fix! 👍
